The `testGlobalDatasetSplitting` unit test has an unexpected error - and likely common among other tests - in python3.
```
[dmwm@189081e99296 WMCore]$ nosetests test/python/WMCore_t/WorkQueue_t/WorkQueue_t.py:WorkQueueTest.testGlobalDatasetSplitting
/home/dmwm/unittestdeploy/wmagent/1.4.9.pre6/sw/slc7_amd64_gcc630/external/condorpy3/8.9.7/lib/python3.8/site-packages/htcondor/__init__.py:48: UserWarning: Neither the environment variable CONDOR_CONFIG, /etc/condor/, /usr/local/etc/, nor ~condor/ contain a condor_config source. Therefore, we are using a null condor_config.
  _warnings.warn(message)
E
======================================================================
ERROR: Dataset splitting at global level
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/WorkQueue_t/WorkQueue_t.py", line 164, in setUp
    self.spec = stepchainWorkload('testProduction', self.mcArgs)
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/WorkQueue_t/WorkQueue_t.py", line 72, in stepchainWorkload
    wmspec = factory.factoryWorkloadConstruction(workloadName, arguments)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMSpec/StdSpecs/StdBase.py", line 887, in factoryWorkloadConstruction
    workload = self.__call__(workloadName=workloadName, arguments=arguments)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMSpec/StdSpecs/StepChain.py", line 99, in __call__
    self.setupGeneratorTask(firstTask, taskConf)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMSpec/StdSpecs/StepChain.py", line 222, in setupGeneratorTask
    outMods = self.setupProcessingTask(task, "Production", couchDBName=self.couchDBName,
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMSpec/StdSpecs/StdBase.py", line 397, in setupProcessingTask
    if 'EventStreams' in taskConf and taskConf['EventStreams'] >= 0:
TypeError: '>=' not supported between instances of 'NoneType' and 'int'
-------------------- >> begin captured stdout << ---------------------
```

Working on it.

--- a/src/python/WMCore/Database/CMSCouch.py
+++ b/src/python/WMCore/Database/CMSCouch.py
@@ -10,7 +10,7 @@ NOT A THREAD SAFE CLASS.
 """
 from __future__ import print_function, division
 from builtins import str as newstr, bytes as newbytes, object
-from Utils.Utilities import decodeBytesToUnicode, encodeUnicodeToBytes
+from Utils.Utilities import decodeBytesToUnicode, encodeUnicodeToBytes, decodeBytesToUnicodeConditional
 from Utils.PythonVersion import PY3
 
 from future import standard_library
@@ -42,6 +42,9 @@ def check_server_url(srvurl):
     good_name = srvurl.startswith('http://') or srvurl.startswith('https://')
     if not good_name:
         raise ValueError('You must include http(s):// in your servers address')
+
+
+PY3_STR_DECODER = lambda x: decodeBytesToUnicodeConditional(x, condition=PY3)
 
 
 class Document(dict):
@@ -353,11 +356,11 @@ class Database(CouchDBRequests):
             updateUri = '/%s/_design/%s/_update/%s/%s?%s' % \
                         (self.name, design, update_func, doc_id, urllib.parse.urlencode(fields))
 
-            return self.put(uri=updateUri, decode=False)
+            return self.put(uri=updateUri, decode=PY3_STR_DECODER)
         else:
             updateUri = '/%s/_design/%s/_update/%s/%s' % \
                         (self.name, design, update_func, doc_id)
-            return self.put(uri=updateUri, data=fields, decode=False)
+            return self.put(uri=updateUri, data=fields, decode=PY3_STR_DECODER)
 
     def updateBulkDocuments(self, doc_ids, paramsToUpdate, updateLimits=1000):
 
@@ -410,7 +413,7 @@ class Database(CouchDBRequests):
         doc_id = urllib.parse.quote_plus(doc_id)
 
         updateUri = '/%s/%s' % (self.name, doc_id)
-        return self.put(uri=updateUri, data=fields, decode=False)
+        return self.put(uri=updateUri, data=fields, decode=PY3_STR_DECODER)
 
     def documentExists(self, id, rev=None):
         """
@@ -570,15 +573,15 @@ class Database(CouchDBRequests):
                 data = urllib.parse.urlencode(encodedOptions)
                 retval = self.post('/%s/_design/%s/_list/%s/%s?%s' % \
                                    (self.name, design, list, view, data), {'keys': keys},
-                                   decode=False)
+                                   decode=PY3_STR_DECODER)
             else:
                 retval = self.post('/%s/_design/%s/_list/%s/%s' % \
                                    (self.name, design, list, view), {'keys': keys},
-                                   decode=False)
+                                   decode=PY3_STR_DECODER)
         else:
             retval = self.get('/%s/_design/%s/_list/%s/%s' % \
                               (self.name, design, list, view), encodedOptions,
-                              decode=False)
+                              decode=PY3_STR_DECODER)
 
         return retval
 
@@ -662,7 +665,7 @@ class Database(CouchDBRequests):
         Retrieve an attachment for a couch document.
         """
         url = "/%s/%s/%s" % (self.name, id, name)
-        attachment = self.get(url, None, encode=False, decode=False)
+        attachment = self.get(url, None, encode=False, decode=PY3_STR_DECODER)
 
         # there has to be a better way to do this but if we're not de-jsoning
         # the return values, then this is all I can do for error checking,

--- a/src/python/WMCore/Services/Requests.py
+++ b/src/python/WMCore/Services/Requests.py
@@ -300,13 +300,13 @@ class Requests(dict):
     def decodeResult(self, result, decoder):
         """
         Decode the http/pycurl request result
+        NOTE: if decoder is provided with a False value, then it means no
+        decoding is applied on the results at all
         """
         if isinstance(decoder, (types.MethodType, types.FunctionType)):
             result = decoder(result)
         elif decoder is not False:
             result = self.decode(result)
-        if PY3:
-            result = decodeBytesToUnicode(result)
         return result
 
     def encode(self, data):
@@ -319,6 +319,8 @@ class Requests(dict):
         """
         decode data to some appropriate format, for now make it a string...
         """
+        if PY3:
+            return decodeBytesToUnicode(data)
         return data.__str__()
 
     def cachePath(self, given_path, service_name):

--- a/src/python/WMCore/WMSpec/Persistency.py
+++ b/src/python/WMCore/WMSpec/Persistency.py
@@ -74,11 +74,8 @@ class PersistencyHelper(object):
             # use own request class so we get authentication if needed
             from WMCore.Services.Requests import Requests
             request = Requests(filename)
-            data = request.makeRequest('', incoming_headers={"Accept": "*/*"})
+            data = request.makeRequest('', incoming_headers={"Accept": "*/*"}, decoder=False)
             self.data = pickle.loads(data[0])
-
-        # TODO: use different encoding scheme for different extension
-        # extension = filename.split(".")[-1].lower()
 
         return
 

--- a/src/python/WMCore/WMSpec/StdSpecs/StdBase.py
+++ b/src/python/WMCore/WMSpec/StdSpecs/StdBase.py
@@ -394,7 +394,7 @@ class StdBase(object):
         eventStreams = self.eventStreams
         if 'Multicore' in taskConf and taskConf['Multicore'] > 0:
             multicore = taskConf['Multicore']
-        if 'EventStreams' in taskConf and taskConf['EventStreams'] >= 0:
+        if taskConf.get("EventStreams") is not None and taskConf['EventStreams'] >= 0:
             eventStreams = taskConf['EventStreams']
         procTaskCmsswHelper.setNumberOfCores(multicore, eventStreams)
 
@@ -1051,7 +1051,7 @@ class StdBase(object):
                      "ProcessingVersion": {"default": 1, "type": int, "validate": procversion},
                      "Memory": {"default": 2300.0, "type": float, "validate": lambda x: x > 0},
                      "Multicore": {"default": 1, "type": int, "validate": lambda x: x > 0},
-                     "EventStreams": {"type": int, "validate": lambda x: x >= 0, "null": True},
+                     "EventStreams": {"type": int, "default": 0, "validate": lambda x: x >= 0, "null": True},
                      "MergedLFNBase": {"default": "/store/data"},
                      "UnmergedLFNBase": {"default": "/store/unmerged"},
                      "DeleteFromSource": {"default": False, "type": strToBool},
@@ -1098,7 +1098,7 @@ class StdBase(object):
                      "ProcessingVersion": {"type": int, "validate": procversion, "assign_optional": True},
                      "Memory": {"type": float, "validate": lambda x: x > 0},
                      "Multicore": {"type": int, "validate": lambda x: x > 0},
-                     "EventStreams": {"type": int, "validate": lambda x: x >= 0, "null": True},
+                     "EventStreams": {"null": True, "default": 0, "type": int, "validate": lambda x: x >= 0},
 
                      "SiteBlacklist": {"default": [], "type": makeList,
                                        "validate": lambda x: all([cmsname(y) for y in x])},
@@ -1180,6 +1180,7 @@ class StdBase(object):
                      'ConfigCacheID': {'optional': False, 'type': str},
                      'DataPileup': {'default': None, 'null': False, 'optional': True, 'type': str,
                                     'validate': dataset},
+                     # for task/step level, the default EventStreams value (0) comes from the top level
                      'EventStreams': {'null': True, 'type': int, 'validate': lambda x: x >= 0},
                      'EventsPerJob': {'null': True, 'type': int, 'validate': lambda x: x > 0},
                      'EventsPerLumi': {'default': None, 'null': True, 'optional': True, 'type': int,

--- a/src/python/WMCore/WMSpec/StdSpecs/StepChain.py
+++ b/src/python/WMCore/WMSpec/StdSpecs/StepChain.py
@@ -321,7 +321,7 @@ class StepChainWorkloadFactory(StdBase):
             eventStreams = self.eventStreams
             if taskConf['Multicore'] > 0:
                 multicore = taskConf['Multicore']
-            if taskConf.get('EventStreams') >= 0:
+            if taskConf.get('EventStreams') > 0:
                 eventStreams = taskConf['EventStreams']
 
             currentCmsswStepHelper.setNumberOfCores(multicore, eventStreams)

--- a/src/python/WMCore/WorkQueue/WMBSHelper.py
+++ b/src/python/WMCore/WorkQueue/WMBSHelper.py
@@ -32,13 +32,12 @@ from WMCore.WMRuntime.SandboxCreator import SandboxCreator
 
 
 def wmbsSubscriptionStatus(logger, dbi, conn, transaction):
-    """Function to return status of wmbs subscriptions
     """
-    action = DAOFactory(package='WMBS',
-                        logger=logger,
-                        dbinterface=dbi)('Monitoring.SubscriptionStatus')
-    return action.execute(conn=conn,
-                          transaction=transaction)
+    Function to return status of wmbs subscriptions
+    """
+    daoFactory = DAOFactory(package='WMCore.WMBS', logger=logger, dbinterface=dbi)
+    action = daoFactory(classname='Monitoring.SubscriptionStatus')
+    return action.execute(conn=conn, transaction=transaction)
 
 
 class WorkQueueWMBSException(WMException):

--- a/src/python/WMQuality/Emulators/DBSClient/MockDbsApi.py
+++ b/src/python/WMQuality/Emulators/DBSClient/MockDbsApi.py
@@ -16,7 +16,7 @@ from RestClient.ErrorHandling.RestClientExceptions import HTTPError
 from WMCore.Services.DBS.DBSErrors import DBSReaderError
 from WMCore.WMBase import getTestBase
 
-from Utils.Utilities import encodeUnicodeToBytes
+from Utils.Utilities import encodeUnicodeToBytesConditional
 from Utils.PythonVersion import PY2
 
 # Read in the data just once so that we don't have to do it for every test (in __init__)
@@ -66,7 +66,7 @@ class MockDbsApi(object):
             origArgs = copy.deepcopy(kwargs)
             returnDicts = []
             for lfn in kwargs['logical_file_name']:
-                origArgs.update({'logical_file_name': [str(lfn)]})
+                origArgs.update({'logical_file_name': [lfn]})
                 returnDicts.extend(self.genericLookup(**origArgs))
             return returnDicts
         else:
@@ -89,7 +89,7 @@ class MockDbsApi(object):
             origArgs = copy.deepcopy(kwargs)
             returnDicts = []
             for lfn in kwargs['logical_file_name']:
-                origArgs.update({'logical_file_name': [str(lfn)]})
+                origArgs.update({'logical_file_name': [lfn]})
                 returnDicts.extend(self.genericLookup(**origArgs))
             return returnDicts
         else:
@@ -114,14 +114,15 @@ class MockDbsApi(object):
         :param kwargs: named arguments it was called with
         :return: the dictionary that DBS would have returned
         """
-
         if self.url not in mockData:
             raise DBSReaderError("Mock DBS emulator knows nothing about instance %s" % self.url)
 
         if kwargs:
-            if PY2:
-                for k in kwargs:
-                    kwargs[k] = encodeUnicodeToBytes(kwargs[k])
+            for k in kwargs:
+                if isinstance(kwargs[k], (list, tuple)):
+                    kwargs[k] = [encodeUnicodeToBytesConditional(item, condition=PY2) for item in kwargs[k]]
+                else:
+                    kwargs[k] = encodeUnicodeToBytesConditional(kwargs[k], condition=PY2)
             signature = '%s:%s' % (self.item, sorted(viewitems(kwargs)))
         else:
             signature = self.item

--- a/test/python/WMCore_t/Services_t/Requests_t.py
+++ b/test/python/WMCore_t/Services_t/Requests_t.py
@@ -1,11 +1,10 @@
 #!/usr/bin/env python
 # -*- coding: ISO-8859-1 -*-
-'''n
+"""
 Created on Aug 6, 2009
 
 @author: meloam
-'''
-
+"""
 from __future__ import print_function
 from future import standard_library
 standard_library.install_aliases()
@@ -120,7 +119,7 @@ class testRepeatCalls(RESTBaseUnitTest):
             time.sleep(i)
             print('test %s starting at %s' % (i, time.time()))
             try:
-                result = req.get('/', incoming_headers={'Cache-Control': 'no-cache'}, decode=False)
+                result = req.get('/', incoming_headers={'Cache-Control': 'no-cache'}, decode=True)
                 self.assertEqual(False, result[3])
                 self.assertEqual(200, result[1])
             except HTTPException as he:
@@ -154,11 +153,11 @@ class testRepeatCalls(RESTBaseUnitTest):
         idict = {'req_cache_path': self.cache_path, 'pycurl': 1}
         req = Requests.Requests(self.urlbase, idict)
         headers = {'Cache-Control': 'no-cache'}
-        self.assertRaises(pycurl.error, req.get, '/', incoming_headers=headers, decode=False)
+        self.assertRaises(pycurl.error, req.get, '/', incoming_headers=headers, decode=True)
 
         # now restart server and hope we can connect
         self.rt.start(blocking=False)
-        result = req.get('/', incoming_headers=headers, decode=False)
+        result = req.get('/', incoming_headers=headers, decode=True)
         self.assertEqual(result[3], False)
         self.assertEqual(result[1], 200)
 
@@ -296,7 +295,7 @@ class TestRequests(unittest.TestCase):
         if not isinstance(json.loads(out[0]), dict):
             msg = 'wrong data type'
             raise Exception(msg)
-        out = req.makeRequest('/auth/trouble', decoder=False)
+        out = req.makeRequest('/auth/trouble', decoder=True)
         self.assertEqual(out[1], 200)
         self.assertNotEqual(out[0].find('passed basic validation'), -1)
         self.assertNotEqual(out[0].find('certificate is a proxy'), -1)
@@ -312,7 +311,7 @@ class TestRequests(unittest.TestCase):
     def testSecureNoAuth_with_pycurl(self):
         """https with no client authentication"""
         req = Requests.Requests('https://cmsweb.cern.ch', {'pycurl': 1})
-        out = req.makeRequest('', decoder=False)
+        out = req.makeRequest('', decoder=True)
         self.assertEqual(out[1], 200)
         # we should get an html page in response
         self.assertNotEqual(out[0].find('html'), -1)
@@ -341,14 +340,14 @@ class TestRequests(unittest.TestCase):
         os.environ.pop('X509_USER_CERT', None)
         os.environ.pop('X509_USER_KEY', None)
         req = Requests.Requests('https://cmsweb.cern.ch:443', {'pycurl': 1})
-        out = req.makeRequest('/auth/trouble', decoder=False)
+        out = req.makeRequest('/auth/trouble', decoder=True)
         self.assertEqual(out[1], 200)
         self.assertNotEqual(out[0].find('passed basic validation'), -1)
 
     def testNoCache(self):
         """Cache disabled"""
         req = Requests.Requests('https://cmssdt.cern.ch/SDT/', {'cachepath': None})
-        out = req.makeRequest('/', decoder=False)
+        out = req.makeRequest('/', decoder=True)
         self.assertEqual(out[3], False)
         self.assertTrue('html' in out[0])
 

--- a/test/python/WMCore_t/WMSpec_t/StdSpecs_t/TaskChain_t.py
+++ b/test/python/WMCore_t/WMSpec_t/StdSpecs_t/TaskChain_t.py
@@ -1057,7 +1057,7 @@ class TaskChainTests(EmulatedUnitTestCase):
         processorDocs = makeProcessingConfigs(self.configDatabase)
 
         arguments = TaskChainWorkloadFactory.getTestArguments()
-        arguments.update(REQUEST_INPUT)
+        arguments.update(deepcopy(REQUEST_INPUT))
         arguments['Task1']['ConfigCacheID'] = processorDocs['DigiHLT']
         arguments['Task2']['ConfigCacheID'] = processorDocs['Reco']
 
@@ -1088,7 +1088,7 @@ class TaskChainTests(EmulatedUnitTestCase):
         processorDocs = makeProcessingConfigs(self.configDatabase)
 
         testArguments = TaskChainWorkloadFactory.getTestArguments()
-        testArguments.update(REQUEST_INPUT)
+        testArguments.update(deepcopy(REQUEST_INPUT))
         testArguments['Task1']['ConfigCacheID'] = processorDocs['DigiHLT']
         testArguments['Task2']['ConfigCacheID'] = processorDocs['Reco']
 
@@ -2297,7 +2297,7 @@ class TaskChainTests(EmulatedUnitTestCase):
         processorDocs = makeProcessingConfigs(self.configDatabase)
 
         arguments = TaskChainWorkloadFactory.getTestArguments()
-        arguments.update(REQUEST_INPUT)
+        arguments.update(deepcopy(REQUEST_INPUT))
         arguments['Task1']['ConfigCacheID'] = processorDocs['DigiHLT']
         arguments['Task2']['ConfigCacheID'] = processorDocs['Reco']
         arguments['Task2']['KeepOutput'] = False

--- a/test/python/WMCore_t/WorkQueue_t/WorkQueue_t.py
+++ b/test/python/WMCore_t/WorkQueue_t/WorkQueue_t.py
@@ -822,25 +822,31 @@ class WorkQueueTest(WorkQueueTestCase):
 
         # queue work, globally for block, pass down, report back -> complete
         totalSpec = 1
-        dummyTotalBlocks = totalSpec * NBLOCKS_HICOMM
         self.assertEqual(0, len(self.globalQueue))
         for _ in range(totalSpec):
             self.globalQueue.queueWork(dqmWorkload.specUrl())
+        # now we have elements inserted in workqueue
+        self.assertEqual(len(self.globalQueue.status(status='Available')), totalSpec)
+        # this should be a continuous processing of the spec, so no changes
         self.globalQueue.processInboundWork()
         self.assertEqual(totalSpec, len(self.globalQueue))
+        self.assertEqual(len(self.globalQueue.status(status='Available')), totalSpec)
 
         # pull to local
         # self.globalQueue.updateLocationInfo()
         self.assertEqual(self.localQueue.pullWork({'T2_XX_SiteA': 1000}), totalSpec)
         syncQueues(self.localQueue)
         self.assertEqual(len(self.localQueue.status(status='Available')), totalSpec)
+        self.assertEqual(len(self.globalQueue.status(status='Acquired')), totalSpec)
         self.localQueue.updateLocationInfo()
         work = self.localQueue.getWork({'T2_XX_SiteA': 1000}, {})
-        self.assertEqual(len(work), 0)
+        self.assertEqual(len(work), 1)
+        self.assertEqual(len(self.localQueue.status(status='Running')), totalSpec)
+        self.assertEqual(len(self.globalQueue.status(status='Acquired')), totalSpec)
         self.localQueue.doneWork([str(x.id) for x in work])
         self.assertEqual(len(self.localQueue.status(status='Done')), totalSpec)
         syncQueues(self.localQueue)
-        # elements are not deleted untill request status is changed
+        # elements are not deleted until request status is changed
         self.assertEqual(len(self.localQueue.status(status='Done')), totalSpec)
         self.assertEqual(len(self.globalQueue.status(status='Done')), totalSpec)
 

